git/gogit/git_client.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gogit
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ProtonMail/go-crypto/openpgp"
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-billy/v5/osfs"
+	extgogit "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/cache"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/filesystem"
+
+	"github.com/fluxcd/pkg/git"
+)
+
+// GoGitClient implements git.GitClient
+type GoGitClient struct {
+	path       string
+	repository *extgogit.Repository
+	authOpts   *git.AuthOptions
+	// useMemFS tells the client to cache all files in memory.
+	useMemFS bool
+}
+
+var _ git.GitClient = &GoGitClient{}
+
+// NewGoGitClient returns a new GoGitClient
+func NewGoGitClient(path string, authOpts *git.AuthOptions, useMemFS bool) *GoGitClient {
+	return &GoGitClient{
+		path:     path,
+		authOpts: authOpts,
+		useMemFS: useMemFS,
+	}
+}
+
+func (g *GoGitClient) Init(ctx context.Context, url, branch string) error {
+	if g.repository != nil {
+		return nil
+	}
+
+	var r *extgogit.Repository
+	var err error
+	if g.useMemFS {
+		storage := filesystem.NewStorage(osfs.New(g.path), cache.NewObjectLRUDefault())
+		r, err = extgogit.Init(storage, memfs.New())
+	} else {
+		r, err = extgogit.PlainInit(g.path, false)
+	}
```

Comment:
It would be great if instead of making this decisions here, we relied on the `Storer` interface and used the same implementation regardless of whether the operation was against a filesystem or memory.